src/Workspaces/Core/Portable/Workspace/Solution/ProjectDependencyGraph_RemoveProject.cs

```diff
@@ -42,74 +42,103 @@ internal ProjectDependencyGraph WithProjectRemoved(ProjectId projectId)
+                foreach (var (id, references) in existingForwardReferencesMap)
                 {
-                    builder[id] = references.Remove(projectId);
+                    var forwardReferences = references.Remove(removedProjectId);
+                    if (forwardReferences.IsEmpty)
+                    {
+                        builder.Remove(id);
+                    }
+                    else
+                    {
+                        builder[id] = forwardReferences;
+                    }
                 }
             }
 
             // Finally, remove 'projectId' itself.
-            builder.Remove(projectId);
+            builder.Remove(removedProjectId);
             return builder.ToImmutable();
         }
 
         /// <summary>
         /// Computes a new <see cref="_lazyReverseReferencesMap"/> for the removal of a project.
         /// </summary>
+        /// <param name="existingReverseReferencesMap">The <see cref="_lazyReverseReferencesMap"/> prior to the removal,
+        /// or <see langword="null"/> if the value prior to removal was not computed for the graph.</param>
         private static ImmutableDictionary<ProjectId, ImmutableHashSet<ProjectId>>? ComputeNewReverseReferencesMapForRemovedProject(
             ImmutableDictionary<ProjectId, ImmutableHashSet<ProjectId>> existingForwardReferencesMap,
             ImmutableDictionary<ProjectId, ImmutableHashSet<ProjectId>>? existingReverseReferencesMap,
-            ProjectId projectId)
+            ProjectId removedProjectId)
         {
             if (existingReverseReferencesMap is null)
             {
                 // The map was never calculated for the previous graph, so there is nothing to update.
                 return null;
             }
 
+            if (!existingForwardReferencesMap.TryGetValue(removedProjectId, out var forwardReferences))
+            {
+                // The removed project did not reference any other projects, so we simply remove it.
+                return existingReverseReferencesMap.Remove(removedProjectId);
+            }
+
             var builder = existingReverseReferencesMap.ToBuilder();
 
-            // Iterate over each project referenced by 'projectId', which is now being removed. Update the reverse
-            // references map for the project to no longer include 'projectId' in the list.
-            if (existingForwardReferencesMap.TryGetValue(projectId, out var referencedProjectIds))
+            // Iterate over each project referenced by 'removedProjectId', which is now being removed. Update the
+            // reverse references map for the project to no longer include 'removedProjectId' in the list.
+            foreach (var referencedProjectId in forwardReferences)
             {
-                foreach (var referencedProjectId in referencedProjectIds)
+                var reverseReferences = existingReverseReferencesMap[referencedProjectId].Remove(removedProjectId);
```

Comment:
Worth extracting this "if empty, remove key else remove from value"?

Reply:
➡️ Included in #41288